File: csharp/circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    private Queue<T> _buff = new Queue<T>();
    private readonly int _capacity;
    
    public CircularBuffer(int capacity) => _capacity = capacity;

    public T Read()
    {
        if (_buff.Count == 0) throw new InvalidOperationException("buffer is empty");

        return _buff.Dequeue();
    }

    public void Write(T value)
    {
        if (_buff.Count == _capacity) throw new InvalidOperationException("buffer is full");
        _buff.Enqueue(value);
    }

    public void Overwrite(T value)
    {
        if (_buff.Count == _capacity) _buff.Dequeue();
        Write(value);
    }

    public void Clear() => _buff.Clear();
}

File: python/circular-buffer/circular_buffer.py
import queue


class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


class CircularBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stored_count = 0
        self.pointer = None
        self.buffer = queue.Queue(capacity)

    def read(self):
        if self.buffer.empty():
            raise BufferEmptyException("Circular buffer is empty")
        return self.buffer.get()

    def write(self, data):
        if self.buffer.full():
            raise BufferFullException("Circular buffer is full")
        self.buffer.put(data)

    def overwrite(self, data):
        if self.buffer.full():
            _ = self.buffer.get()
        self.buffer.put(data)

    def clear(self):
        while not self.buffer.empty():
            _ = self.buffer.get()
